lucene/sandbox/src/java/org/apache/lucene/search/IndexSortSortedNumericDocValuesRangeQuery.java

```diff
@@ -0,0 +1,305 @@
+   * every time the binary search accesses an earlier element.
+   *
+   * We must also account for missing values when performing the binary search. For this
+   * reason, we load the {@link FieldComparator} instead of checking the docvalues directly.
+   * The returned {@link DocIdSetIterator} makes sure to wrap the original docvalues to skip
+   * over documents with no value.
+   */
+  private DocIdSetIterator getDocIdSetIterator(SortField sortField,
+                                               LeafReaderContext context,
+                                               DocIdSetIterator delegate) throws IOException {
+    int maxDoc = context.reader().maxDoc();
+    if (maxDoc <= 0) {
+      return DocIdSetIterator.empty();
+    }
+
+    long lower = sortField.getReverse() ? upperValue : lowerValue;
+    long upper = sortField.getReverse() ? lowerValue : upperValue;
+
+    // Perform a binary search to find the first document with value >= lower.
+    ValueComparator comparator = loadComparator(sortField, lower, context);
+    int low = 0;
+    int high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) <= 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, lower, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+    int firstDocIdInclusive = high + 1;
+
+    // Perform a binary search to find the first document with value > upper.
+    // Since we know that upper >= lower, we can initialize the lower bound
+    // of the binary search to the result of the previous search.
+    comparator = loadComparator(sortField, upper, context);
+    low = firstDocIdInclusive;
+    high = maxDoc - 1;
+
+    while (low <= high) {
+      int mid = (low + high) >>> 1;
+      if (comparator.compare(mid) < 0) {
+        high = mid - 1;
+        comparator = loadComparator(sortField, upper, context);
+      } else {
+        low = mid + 1;
+      }
+    }
+
+    int lastDocIdExclusive = high + 1;
+    return new BoundedDocSetIdIterator(firstDocIdInclusive, lastDocIdExclusive, delegate);
+  }
+
+  /**
+   * Compares the given document's value with a stored reference value.
+   */
+  private interface ValueComparator {
+    int compare(int docID) throws IOException;
+  }
+
+  private static ValueComparator loadComparator(SortField sortField,
+                                                long topValue,
+                                                LeafReaderContext context) throws IOException {
+    @SuppressWarnings("unchecked")
+    FieldComparator<Long> fieldComparator = (FieldComparator<Long>) sortField.getComparator(1, 0);
+    fieldComparator.setTopValue(topValue);
+
+    LeafFieldComparator leafFieldComparator = fieldComparator.getLeafComparator(context);
+    int direction = sortField.getReverse() ? -1 : 1;
+
+    return doc -> {
+      int value = leafFieldComparator.compareTop(doc);
+      return direction * value;
+    };
```

Comment:
I think we need to do the comparison manually instead of using the comparator because of missing values. The comparator will consider documents that don't have a value and documents that have the missing value equal, which doesn't work for us.